Fix various gtkdoc issues

thiblahute opened the pull request:
This patchset is based on top of #739 as it would otherwise conflict too much.

It fixes usage of a main module SGML file that defines the structure of the documentation, until now, that file was not taken into account and gtkdoc was generating one instead,

jpakkane commented:
#739 is now in but this conflicts. Probably just a question of rebase.

thiblahute commented:
Updated.

jpakkane commented:
Since gtk-doc 1.25 has landed to Fedora, Ubuntu and Debian, maybe we should re-enable the gtkdoc tests. Could you enable it and run the tests (and possibly update it if this changes gtkdoc output)? Thanks.

thiblahute commented:
> Since gtk-doc 1.25 has landed to Fedora, Ubuntu and Debian, maybe we should re-enable the gtkdoc tests. Could you enable it and run the tests (and possibly update it if this changes gtkdoc output)?

It is still broken here with gtk-doc 1.25 as I am getting `warning: failed to load external entity "cases/test/10"` (and with master too) so I have the impression the fix was no enough :/

``` 
INSERT  thiblahute  ~  1  gtkdoc-mkhtml --version
1.25
```

I hacked around it to check if it works with that new branch (and master) and with that uter ugly 'patch' it works (so it does on master):

```diff
commit 6ecf1b331d365f4bc8cbc3a6a62e7b792406deff
Author: Thibault Saunier <tsaunier@gnome.org>
Date:   Tue Sep 27 17:11:19 2016 -0300

    Uter ugly 'patch' to get gtkdoc test working

diff --git a/run_tests.py b/run_tests.py
index b57dd39..14dc17d 100755
--- a/run_tests.py
+++ b/run_tests.py
@@ -301,7 +301,7 @@ def _run_test(testdir, test_build_dir, install_dir, extra_args, flags, compile_c
 
 def gather_tests(testdir):
     tests = [t.replace('\\', '/').split('/', 2)[2] for t in glob(os.path.join(testdir, '*'))]
-    testlist = [(int(t.split()[0]), t) for t in tests]
+    testlist = [(int(t.split('_')[0]), t) for t in tests]
     testlist.sort()
     tests = [os.path.join(testdir, t[1]) for t in testlist]
     return tests
@@ -319,23 +319,7 @@ def have_d_compiler():
 
 def detect_tests_to_run():
     all_tests = []
-    all_tests.append(('common', gather_tests('test cases/common'), False))
-    all_tests.append(('failing', gather_tests('test cases/failing'), False))
-    all_tests.append(('prebuilt', gather_tests('test cases/prebuilt'), False))
-
-    all_tests.append(('platform-osx', gather_tests('test cases/osx'), False if mesonlib.is_osx() else True))
-    all_tests.append(('platform-windows', gather_tests('test cases/windows'), False if mesonlib.is_windows() else True))
-    all_tests.append(('platform-linux', gather_tests('test cases/linuxlike'), False if not (mesonlib.is_osx() or mesonlib.is_windows()) else True))
-    all_tests.append(('framework', gather_tests('test cases/frameworks'), False if not mesonlib.is_osx() and not mesonlib.is_windows() else True))
-    all_tests.append(('java', gather_tests('test cases/java'), False if not mesonlib.is_osx() and shutil.which('javac') else True))
-    all_tests.append(('C#', gather_tests('test cases/csharp'), False if shutil.which('mcs') else True))
-    all_tests.append(('vala', gather_tests('test cases/vala'), False if shutil.which('valac') else True))
-    all_tests.append(('rust', gather_tests('test cases/rust'), False if shutil.which('rustc') else True))
-    all_tests.append(('d', gather_tests('test cases/d'), False if have_d_compiler() else True))
-    all_tests.append(('objective c', gather_tests('test cases/objc'), False if not mesonlib.is_windows() else True))
-    all_tests.append(('fortran', gather_tests('test cases/fortran'), False if shutil.which('gfortran') else True))
-    all_tests.append(('swift', gather_tests('test cases/swift'), False if shutil.which('swiftc') else True))
-    all_tests.append(('python3', gather_tests('test cases/python3'), False if shutil.which('python3') else True))
+    all_tests.append(('framework', gather_tests('test_cases/test'), False if not mesonlib.is_osx() and not mesonlib.is_windows() else True))
     return all_tests
 
 def run_tests(extra_args):
@@ -359,7 +343,7 @@ def run_tests(extra_args):
         for t in test_cases:
             # Jenkins screws us over by automatically sorting test cases by name
             # and getting it wrong by not doing logical number sorting.
-            (testnum, testbase) = os.path.split(t)[-1].split(' ', 1)
+            (testnum, testbase) = os.path.split(t)[-1].split('_', 1)
             testname = '%.3d %s' % (int(testnum), testbase)
             result = executor.submit(run_test, skipped, t, extra_args, unity_flags + backend_flags, compile_commands, install_commands, name != 'failing')
             futures.append((testname, t, result))
diff --git a/test cases/frameworks/10 gtk-doc/installed_files.txt.bak b/test cases/frameworks/10 gtk-doc/installed_files.txt
similarity index 100%
rename from test cases/frameworks/10 gtk-doc/installed_files.txt.bak
rename to test cases/frameworks/10 gtk-doc/installed_files.txt
diff --git a/test cases/frameworks/10 gtk-doc/meson.build b/test cases/frameworks/10 gtk-doc/meson.build
index c6881ab..13cbb6b 100644
--- a/test cases/frameworks/10 gtk-doc/meson.build	
+++ b/test cases/frameworks/10 gtk-doc/meson.build	
@@ -8,4 +8,4 @@ inc = include_directories('include')
 
 # We have to disable this test until this bug fix has landed to
 # distros https://bugzilla.gnome.org/show_bug.cgi?id=753145
-# subdir('doc')
+subdir('doc')
diff --git a/test cases/test/10_gtk-doc b/test cases/test/10_gtk-doc
new file mode 120000
index 0000000..5efd0de
--- /dev/null
+++ b/test cases/test/10_gtk-doc	
@@ -0,0 +1 @@
+../frameworks/10 gtk-doc
\ No newline at end of file
diff --git a/test_cases b/test_cases
new file mode 120000
index 0000000..6fe9d5c
--- /dev/null
+++ b/test_cases
@@ -0,0 +1 @@
+test cases
\ No newline at end of file
```

We will need to investigate again why it is still failling, also could you give it a try to see if it is something wrong with me? (using archlinux here)

jpakkane commented:
It works on Debian if I run it by hand but the command invocation gives the wrong path to the sgml file (`../foobar-docs.sgml` which should probably be either absolute, `../path/to/subdir/foobar-docs.sgml` or plain `foobar-docs.sgml`, I don't remember offhand what the working dir is).

thiblahute commented:
`../foobar-docs.sgml` should be correct when calling `gtkdoc-mkhtml` as this command is called from `html/` to build the html files in there.

jpakkane commented:
If I run it there it works. If I run it via the command it fails with `'gtkdoc-mkhtml' failed with status 4`.

thiblahute commented:
Can you show me in what case it works exactly?

In my case it always fails and printing stuff in gtkdoc-mkhtml shows me:

```
cd '/home/thiblahute/devel/meson/test cases/frameworks/10 gtk-doc/build/doc/html' && 'gtkdoc-mkhtml' '--path=/home/thiblahute/devel/meson/test cases/frameworks/10 gtk-doc/include' 'foobar' '../foobar-docs.sgml'
Running: /usr/bin/xsltproc --path /home/thiblahute/devel/meson/test cases/frameworks/10 gtk-doc/include --nonet --xinclude --stringparam gtkdoc.bookname foobar --stringparam gtkdoc.version 1.25 /usr/share/gtk-doc/data/gtk-doc.xsl ../foobar-docs.sgml
NOOO! exit 4
 warning: failed to load external entity "cases/frameworks/10"
cannot parse cases/frameworks/10
 ```

Looks similare to https://bugzilla.gnome.org/show_bug.cgi?id=753145 to me.

thiblahute commented:
And:
```
(mesonenv)  ➜  html git:(mymaster) ✗  /usr/bin/xsltproc --path "/home/thiblahute/devel/meson/test cases/frameworks/10 gtk-doc/include" --nonet --xinclude --stringparam gtkdoc.bookname foobar --stringparam gtkdoc.version 1.25 /usr/share/gtk-doc/data/gtk-doc.xsl ../foobar-docs.sgml
Computing chunks...
Writing foobar-foo.html for refentry(foobar-foo)
Writing foobar.html for reference(foobar)
Writing index.html for book(index)
Writing foobar.devhelp2 for book(index)
```

That works.

thiblahute commented:
I proposed a patch for https://bugzilla.gnome.org/show_bug.cgi?id=753145

jpakkane commented:
So no enabling the tests then. But I trust you have tested this branch on real world code?

thiblahute commented:
I did yes, GStreamer documentation (apart from plugins) is now properly generated.

nirbheek commented:
The relevant bug report for the spaces-in-search-path issue is https://bugzilla.gnome.org/show_bug.cgi?id=769125. It's been ignored for a while, and I don't know what we can do to get it in.
